src/Features/CSharp/Portable/CodeRefactorings/ConvertToRecord/ConvertToRecordHelpers.cs

```diff
@@ -0,0 +1,914 @@
+                                    }
+                                }
+                            }
+                        }]
+                    }]
+                } && referencedParameter1.Equals(parameter))
+            {
+                assignedVariable = castOther;
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Get the referenced parameter (and unwraps implicit cast if necessary) or null if a parameter wasn't referenced
+        /// </summary>
+        /// <param name="operation">The operation for which to get the parameter</param>
+        /// <returns>the referenced parameter or null if unable to find</returns>
+        private static IParameterSymbol? GetParamFromArgument(IOperation operation)
+            => (operation.WalkDownConversion() as IParameterReferenceOperation)?.Parameter;
+
+        private static ISymbol? GetReferencedSymbolObject(IOperation reference)
+        {
+            return reference.WalkDownConversion() switch
+            {
+                IInstanceReferenceOperation thisReference => thisReference.Type,
+                ILocalReferenceOperation localReference => localReference.Local,
+                IParameterReferenceOperation paramReference => paramReference.Parameter,
+                _ => null,
+            };
+        }
+
+        // matches form
+        // c1.Equals(c2)
+        // where c1 and c2 are parameter references
+        private static bool IsDotEqualsInvocation(IOperation operation)
+        {
+            // must be called on one of the parameters
+            if (operation is not IInvocationOperation
+                {
+                    TargetMethod.Name: nameof(Equals),
+                    Instance: IOperation instance,
+                    Arguments: [IArgumentOperation { Value: IOperation arg }]
+                })
+            {
+                return false;
+            }
+
+            // get the (potential) parameters, uwrapping any potential implicit casts
+            var invokedOn = GetParamFromArgument(instance);
+            var param = GetParamFromArgument(arg);
+            // make sure we're not referencing the same parameter twice
+            return param != null && invokedOn != null && !invokedOn.Equals(param);
+        }
+
+        // checks for binary expressions of the type otherC == null or null == otherC
+        // or a pattern against null like otherC is (not) null
+        // and "otherC" is a reference to otherObject
+        // takes successRequirement so that if we're in a constext where the operation evaluating to true
+        // would end up being false within the equals method, we look for != instead
+        private static bool IsNullCheck(
+            IOperation operation,
+            bool successRequirement,
+            ISymbol otherObject)
+        {
+            return operation switch
```

Comment:
This pattern seems pretty overkill in this case; can we just write this as:

```
if (operation is IBinaryOperation binaryOperation)
{
}
else if (operation is IIsPatternOperation patternOperation)
{
    if (successRequirement) ...
}
```

That way you can switch on the successRequirement stuff a bit more simply?